src/py21cmfast/wrapper.py

```diff
@@ -2211,17 +2249,31 @@ def run_coeval(
             singleton = True
             redshift = [redshift]
 
-        if not perturb:
-            for z in redshift:
-                perturb += [
-                    perturb_field(
-                        redshift=z,
-                        init_boxes=init_box,
-                        regenerate=regenerate,
-                        write=write,
-                        direc=direc,
-                    )
-                ]
+        # Get the list of redshift we need to scroll through.
+        redshifts = _get_redshifts(flag_options, redshift)
+
+        # Get all the perturb boxes early. We need to get the perturb at every
+        # redshift, even if we are interpolating the perturb field, because the
+        # ionize box needs it.
+        pz = [p.redshift for p in perturb]
+        perturb = [
+            perturb_field(
+                redshift=z,
+                init_boxes=init_box,
+                regenerate=regenerate,
+                hooks=hooks,
+                direc=direc,
+            )
+            if z not in pz
+            else perturb[pz.index(z)]
```

Comment:
This to me seems like all perturb fields are held in memory at any one time. Is this correct? Why aren't these loaded as required as per `run_lightcone` below?

Reply:
Yes, I forgot to go back and change `run_coeval`, will do that.